dex/src/main/java/io/crate/data/BatchIterator.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data;
+
+import java.util.concurrent.CompletionStage;
+
+
+/**
+ * A iterator used to navigate over batched data.
+ * A iterator usually contains some data already in-memory.
+ * This data can be accessed via {@link #moveNext()} and {@link #currentRow()}.
+ *
+ * Once all in-memory data has been consumed more data can loaded with {@link #loadNextBatch()}
+ * unless {@link #allLoaded()} is true in which case the iterator is exhausted.
+ *
+ * A BatchIterator starts either *before* the first row or in an unloaded state.
+ * This means a consumer can consume a BatchIterator like this:
+ *
+ * <pre>
+ *     while (it.moveNext()) {
+ *         // do something with the row
+ *     }
+ *     if (it.allLoaded()) {
+ *          // iterator is exhausted
+ *     } else {
+ *         it.loadNextBatch().whenComplete((r, t) -> {
+ *             // continue consumption
+ *         }
+ *     }
+ * </pre>
+ *
+ *
+ * Thread-safety notes:
+ *
+ * Concurrent usage of a BatchIterator is not supported.
+ */
+public interface BatchIterator {
+
+    Row OFF_ROW = new Row() {
+        @Override
+        public int numColumns() {
+            throw new IllegalStateException("Iterator is not on a row");
+        }
+
+        @Override
+        public Object get(int index) {
+            throw new IllegalStateException("Iterator is not on a row");
+        }
+
+        @Override
+        public Object[] materialize() {
+            throw new IllegalStateException("Iterator is not on a row");
+        }
+    };
+
+    /**
+     * Moves the Iterator back to the starting position.
+     *
+     * A consumer can then iterate over the rows again by calling into
+     * {@link #moveNext()} and {@link #loadNextBatch()} as appropriate.
+     *
+     * @throws IllegalStateException if the cursor is closed
+     */
+    void moveToStart();
```

Comment:
Currently all can implement it. We'll extend the interface once we have implementations that can't or where it'd have a performance impact to provide this functionality.